pkg/core/region.go

```diff
@@ -1833,6 +1833,42 @@ func (r *RegionsInfo) ScanRegions(startKey, endKey []byte, limit int) []*RegionI
 	return res
 }
 
+// BatchScanRegions scans regions in given key pairs, returns at most `limit` regions.
+// limit <= 0 means no limit.
+// The given key pairs should be non-overlapping.
+func (r *RegionsInfo) BatchScanRegions(keyRanges *KeyRanges, limit int) []*RegionInfo {
+	r.t.RLock()
+	defer r.t.RUnlock()
+
+	krs := keyRanges.Ranges()
+	res := make([]*RegionInfo, 0, len(krs))
+	var lastRegion *RegionInfo
+	for _, keyRange := range krs {
+		if lastRegion != nil {
+			if lastRegion.Contains(keyRange.EndKey) {
+				continue
+			} else if lastRegion.Contains(keyRange.StartKey) {
+				keyRange.StartKey = lastRegion.GetEndKey()
+			}
+		}
+		if limit > 0 && len(res) >= limit {
+			return res
+		}
```

Comment:
Do we need to change the order?
```suggestion
		if limit > 0 && len(res) >= limit {
			return res
		}
		if lastRegion != nil {
			if lastRegion.Contains(keyRange.EndKey) {
				continue
			} else if lastRegion.Contains(keyRange.StartKey) {
				keyRange.StartKey = lastRegion.GetEndKey()
			}
		}
```

Reply:
It's on purposem, if the last region within limitation can contains more given ranges, we should not return it earlier.

Reply:
I am sorry for not getting the point, why can the order changed to reject `a region within limitation`? 0 0 
- Because this change seems to just return earlier with greater than limitation condition.

Are you expecting the parameter's value to be changed in place?

Reply:
You're right, the ranges contained by the last region should be decided by the client.